### ISSUE TITLE
UniqueIndex not Unique in Parent -> Child relations.

### ISSUE THREAD

**Issue by oddbear** [bchavez/Bogus#57]
If I understand the code and documentation correctly, UniqueIndex should be globaly unique. This is not the case when generating child objects.

Expected result UniqueIndex:
[{0, 1}, {2, 3}, {4, 5}]

Actual result UniqueIndex:
[{0, 0}, {2, 2}, {4, 4}]

```csharp
var childFaker = new Faker<Child>()
   .RuleFor(u => u.Id, f => f.UniqueIndex);

var parentFaker = new Faker<Parent>()
   .RuleFor(u => u.Id, f => f.UniqueIndex)
   .RuleFor(u => u.Child, f => childFaker.Generate());

parentFaker.Generate(3).Select(o => new { o.Id, CId = o.Child.Id }).Dump(); //Dump: LinqPad only
```

However if I generate those individual like:
```csharp
var childFaker = new Faker<Child>()
   .RuleFor(u => u.Id, f => f.UniqueIndex);

var parentFaker = new Faker<Parent>()
   .RuleFor(u => u.Id, f => f.UniqueIndex);

parentFaker.Generate(3).Dump();
childFaker.Generate(3).Dump();
```

The result will be like expected:
[0, 1, 2]
[3, 4, 5]

```csharp
public class Parent
{
   public int Id { get; set; }
   public Child Child { get; set; }
}

public class Child
{
   public int Id { get; set; }
}
```

Edit: This might be related to https://github.com/bchavez/Bogus/issues/48
However the issue itself should different.
_Testet in Bogus from NuGet v10.0.1_

**Comment by bchavez** [comment]
Hello @oddbear ,

Rules are executed in-order of being defined. And `UniqueIndex` is only updated after a `Generation()` call. So, if you flatten the calls, in your example:
```
   executes parent.RuleFor(u => u.Id, f => f.UniqueIndex)
   executes child.RuleFor(u => u.Id, f => f.UniqueIndex);
   child generation done. UniqueIndex++
   parent generation done. UniqueIndex++
```
So, if you swap the order of `UniqueIndex` in the `Parent` it should work.
```
var childFaker = new Faker<Child>()
   .RuleFor(u => u.Id, f => f.UniqueIndex);

var parentFaker = new Faker<Parent>()
   .RuleFor(u => u.Child, f => childFaker.Generate())
   .RuleFor(u => u.Id, f => f.UniqueIndex);
```
I agree with you, this seems like an unexpected behavior and awkward, so we should change it. Maybe we can increment before we apply rules instead. I'll have to take a look. Thanks for reporting. :smile_cat: 

:zap: :runner: ***["I was struck by lighting. Walkin' down the street..."](https://www.youtube.com/watch?v=iypUpv9xelg)***

**Comment by bchavez** [comment]
Hi @oddbear,

This `UniqueIndex` awkwardness should be fixed in [`Bogus v11.0.1`](https://www.nuget.org/packages/Bogus/11.0.1). Thanks for the report.

![screen963](https://cloud.githubusercontent.com/assets/478118/23187980/cd2296ac-f841-11e6-8139-711c99418854.png)

-Brian

:watch: :city_sunset: ***["I just can't wait... I just can't wait... for saturday night..."](https://www.youtube.com/watch?v=wOjGmpmvGz8#t=2m)***
